Adding configuration description for TLS/SASL for the KafkaChannel (#3067)
knative/docs: docs/eventing/samples/kafka/channel/README.md

```
kubectl logs --selector='serving.knative.dev/service=broker-kafka-display' -c user-container
```

## Authentication against an Apache Kafka broker

> The KafkaSource supports TLS and SASL authentication methods.

Reviewer: It's not clear to me from this how this actually works; the title mentions the broker but the description mentions the source. Can we get some more detail about what TLS and SASL mean in this context, why would someone use it, how do the broker and source fit in?

Author: Apache Kafka - the message broker (I will remove the broker, since I am NOT talking Knative Broker here)

Reviewer: I still don't think it's clear what is actually being authenticated against what here, or why

Author: OK, not sure how to better say it.

"If Apache Kafka is running in a secured mode, for instance using TLS or SASL, this section show how to confiugure the channel to work against a protected Apache Kafka cluster" ?

### TLS authentication

> For enabling TLS authentication, please have the below files
> 
> * CA Certificate
> * Client Certificate and Key
> 
> `KafkaChannel` expects these files to be in pem format, if it is in other format like jks, please convert to pem.

Reviewer: ```suggestion

To use TLS authentication you must create:

* A CA certificate
* A client certificate and key

**NOTE:** Kafka channels require these files to be in `.pem` format. If your files are in a different format, you must convert them to `.pem`.
```

1. Create the certificate files as secrets in the namespace of your choice, like:
```
$ kubectl create secret --namespace knative-eventing generic my-tls-secret \
  --from-file=ca.crt=caroot.pem \
  --from-file=user.crt=certificate.pem \
  --from-file=user.key=key.pem

>  ```

Reviewer: Won't let me suggest properly, but I'd make some tweaks here:

1. Create the certificate files as secrets in your chosen namespace:

```
$ kubectl create secret --namespace <namespace> generic <secret-name> \
  --from-file=ca.crt=caroot.pem \
  --from-file=user.crt=certificate.pem \
  --from-file=user.key=key.pem
 ```

Reviewer: @matzew doesn't look like this one was updated, or the one below, e.g. "like..." and "Now all you need is to reference..." are very fluffy, that's why I suggested making these more minimal and fluff-free

Author: I think it was updated:
https://github.com/knative/docs/pull/3067/files#diff-4259c7eeed66d201ebf025c842c68c0ad1a5c75121bde0869b80dc21ed7a14e8R170-R180

OK.. I did not see initially the suggestion about "Create the certificate files as secrets in your chosen namespace:" - will change

*NOTE:* It is important to use the same keys (`ca.crt`, `user.crt` and `user.key`).

Now all you need is to reference your secret and its namespace on the `config-kafka` ConfigMap:
```
apiVersion: v1
kind: ConfigMap
metadata:
  name: config-kafka
  namespace: knative-eventing
data:
  #bootstrapServers: REPLACE_WITH_CLUSTER_URL
  #authSecretName: name-of-your-secret-for-kafka-auth
  #authSecretNamespace: namespace-of-your-secret-for-kafka-auth

>  ```

Reviewer: 1. Reference the secret and its namespace in the `config-kafka` ConfigMap:

```
apiVersion: v1
kind: ConfigMap
metadata:
  name: config-kafka
  namespace: knative-eventing
data:
  #bootstrapServers: REPLACE_WITH_CLUSTER_URL
  #authSecretName: name-of-your-secret-for-kafka-auth
  #authSecretNamespace: namespace-of-your-secret-for-kafka-auth
 ```

### SASL authentication

> For enabling SASL authentication, please have the following data:
> 
> * Username and password
> * Type of use SASL mechanism (e.g. `PLAIN`, `SCRAM-SHA-256` or `SCRAM-SHA-512`)

Reviewer: ```suggestion
To use SASL authentication, you will need the following information:

* A username and password.
* The type of SASL mechanism you wish to use. For example; `PLAIN`, `SCRAM-SHA-256` or `SCRAM-SHA-512`.

**NOTE:** It is recommended to also enable TLS. If you enable this, you will also need the `ca.crt` certificate as described in the previous section.
```

1. Create the certificate files as secrets in the namespace of your choice, like:
```
$ kubectl create secret --namespace knative-eventing generic my-sasl-secret \
  --from-file=ca.crt=caroot.pem \
  --from-literal=password="SecretPassword" \
  --from-literal=saslType="SCRAM-SHA-512" \
  --from-literal=user="my-sasl-user"

> ```

Reviewer: 1. Create the certificate files as secrets in your chosen namespace:

```
$ kubectl create secret --namespace <namespace> generic <secret-name> \
  --from-file=ca.crt=caroot.pem \
  --from-literal=password="SecretPassword" \
  --from-literal=saslType="SCRAM-SHA-512" \
  --from-literal=user="my-sasl-user"
```
```

> *NOTE:* It is important to use the same keys (`user`, `password` and `saslType`). It's also recommended to enable TLS, in which case you need the `ca.crt` key as well! 

Reviewer: ```suggestion
*NOTE:* It is important to use the same keys; `user`, `password` and `saslType`.
```

Reviewer: For the second part of this sentence make it a separate note at the start of this section.

Now all you need is to reference your secret and its namespace on the `config-kafka` ConfigMap:

```
apiVersion: v1
kind: ConfigMap
metadata:
  name: config-kafka
  namespace: knative-eventing
data:

>   #bootstrapServers: REPLACE_WITH_CLUSTER_URL

Reviewer: Can we use a consistent format for variables please? Can we maybe format them all `<like-this>` as per the Kubernetes style guide? https://kubernetes.io/docs/contribute/style/style-guide/#use-angle-brackets-for-placeholders

Reviewer: These still seem inconsistent, @matzew

Author: @abrennan89 I think this section is outdated (see the `Outdated` header above the code, next to the file name)

here is the actual file: https://github.com/knative/docs/pull/3067/files#diff-4259c7eeed66d201ebf025c842c68c0ad1a5c75121bde0869b80dc21ed7a14e8R216-R225

  #authSecretName: name-of-your-secret-for-kafka-auth
  #authSecretNamespace: namespace-of-your-secret-for-kafka-auth

> ```

Reviewer: 1. Reference the secret and its namespace in the `config-kafka` ConfigMap:

```
apiVersion: v1
kind: ConfigMap
metadata:
  name: config-kafka
  namespace: knative-eventing
data:
  #bootstrapServers: <cluster-URL>
  #authSecretName: <name-of-your-secret-for-kafka-auth>
  #authSecretNamespace: <namespace-of-your-secret-for-kafka-auth>
```

```
kubectl logs --selector='serving.knative.dev/service=broker-kafka-display' -c user-container
```

## Authentication against an Apache Kafka

> In production enviornments it is common that the Apache Kafka cluster is secured using [TLS](http://kafka.apache.org/documentation/#security_ssl) or [SASL](http://kafka.apache.org/documentation/#security_sasl). This section show how to confiugure the `KafkaChannel` to work against a protected Apache Kafka cluster, with the two supported TLS and SASL authentication methods.

Reviewer 2: ```suggestion
In production enviornments it is common that the Apache Kafka cluster is secured using [TLS](http://kafka.apache.org/documentation/#security_ssl) or [SASL](http://kafka.apache.org/documentation/#security_sasl). This section shows how to confiugure the `KafkaChannel` to work against a protected Apache Kafka cluster, with the two supported TLS and SASL authentication methods.
```

### TLS authentication

To use TLS authentication you must create:

* A CA certificate
* A client certificate and key

**NOTE:** Kafka channels require these files to be in `.pem` format. If your files are in a different format, you must convert them to `.pem`.

1. Create the certificate files as secrets in your chosen namespace:
```
$ kubectl create secret --namespace <namespace> generic <kafka-auth-secret> \
  --from-file=ca.crt=caroot.pem \
  --from-file=user.crt=certificate.pem \
  --from-file=user.key=key.pem
 ```

*NOTE:* It is important to use the same keys (`ca.crt`, `user.crt` and `user.key`).

> Now all you need is to reference your secret and its namespace on the `config-kafka` ConfigMap:

Reviewer: Same comment as "Now all you need...." in other section below. Make it a step 2 without the fluff.

```
apiVersion: v1
kind: ConfigMap
metadata:
  name: config-kafka
  namespace: knative-eventing
data:

>   bootstrapServers: <bootstrapServers>

Reviewer: `<bootstrap-servers>` for consistency

  authSecretName: <kafka-auth-secret>
  authSecretNamespace: <namespace>
 ```

### SASL authentication

To use SASL authentication, you will need the following information:

* A username and password.
* The type of SASL mechanism you wish to use. For example; `PLAIN`, `SCRAM-SHA-256` or `SCRAM-SHA-512`.

**NOTE:** It is recommended to also enable TLS. If you enable this, you will also need the `ca.crt` certificate as described in the previous section.

1. Create the certificate files as secrets in your chosen namespace:
```
$ kubectl create secret --namespace <namespace> generic <kafka-auth-secret> \
  --from-file=ca.crt=caroot.pem \
  --from-literal=password="SecretPassword" \
  --from-literal=saslType="SCRAM-SHA-512" \
  --from-literal=user="my-sasl-user"
```

*NOTE:* It is important to use the same keys; `user`, `password` and `saslType`.

> Now all you need is to reference your secret and its namespace on the `config-kafka` ConfigMap:

Reviewer 2: ```suggestion
Now all you need is to reference your secret and its namespace in the `config-kafka` ConfigMap:
```

Reviewer: Seems to be still missing a suggestion here, should be a step 2 without the fluff...
e.g.

> 2. Reference your secret and the namespace of the secret in the `config-kafka` ConfigMap:

```
apiVersion: v1
kind: ConfigMap
metadata:
  name: config-kafka
  namespace: knative-eventing
data:

>   bootstrapServers: <bootstrapServers>

Reviewer: `<bootstrap-servers>` for consistency

```
kubectl logs --selector='serving.knative.dev/service=broker-kafka-display' -c user-container
```

> ## Authentication against an Apache Kafka

Reviewer: Authentication of what against Kafka though? Again, I still think this is confusing because underneath in the intro line we only talk about KafkaSource - I would change the heading to something like "Authenticating a `KafkaSource` object using TLS and SASL", or whatever is accurate, that states what is being authenticated by what.

Author: the source was a typo - I am sorry. This doc is only for the channel (Source has already docs for that, since a long time)

I will fix the intro, best I can

Reviewer: This is so much better, thank you so much Matthias! <3

The KafkaSource supports TLS and SASL authentication methods.

### TLS authentication

> 

Reviewer: remove extra line?

To use TLS authentication you must create:

* A CA certificate
* A client certificate and key

**NOTE:** Kafka channels require these files to be in `.pem` format. If your files are in a different format, you must convert them to `.pem`.

> 

Reviewer: not sure why but there seem to be a few parts with unnecessary empty lines, I'd remove them

```
kubectl logs --selector='serving.knative.dev/service=broker-kafka-display' -c user-container
```

## Authentication against an Apache Kafka broker

The KafkaSource supports TLS and SASL authentication methods.

### TLS authentication

For enabling TLS authentication, please have the below files

* CA Certificate
* Client Certificate and Key

`KafkaChannel` expects these files to be in pem format, if it is in other format like jks, please convert to pem.

1. Create the certificate files as secrets in the namespace of your choice, like:

   ```
  $ kubectl create secret --namespace knative-eventing generic my-tls-secret \
    --from-file=ca.crt=caroot.pem \
    --from-file=user.crt=certificate.pem \
    --from-file=user.key=key.pem

>    ```

Reviewer 3: Some formatting issue here. See the rendered HTML

> NOTE: It is important to use the same keys (`ca.crt`, `user.crt` and `user.key`).

Now all you need is to reference your secret and its namespace on the `config-kafka` ConfigMap:

```
apiVersion: v1
kind: ConfigMap
metadata:
  name: config-kafka
  namespace: knative-eventing
data:
  #bootstrapServers: REPLACE_WITH_CLUSTER_URL
  #authSecretName: name-of-your-secret-for-kafka-auth
  #authSecretNamespace: namespace-of-your-secret-for-kafka-auth

```

### SASL authentication

For enabling SASL authentication, please have the following data:

* Username and password

> * Type of use SASL mechanism (e.g. `PLAIN`, `SCRAM-SHA-256` or `SCRAM-SHA-512`)

Reviewer 3: Also CA Certificate

Author: CA file is only needed if running w/ TLS (+SASL)

Reviewer 3: But, what about `--from-file=ca.crt=...` at https://github.com/knative/docs/pull/3067/files#diff-4259c7eeed66d201ebf025c842c68c0ad1a5c75121bde0869b80dc21ed7a14e8R202

Author: that's for TLS w/ SASL - I added the `ca.crt` thing on the NOTE below the actual code

Reviewer: Add the note here instead please, no point telling someone about it halfway through the procedure ;)
